### PR TITLE
feat: metrics for sqlx conn pools

### PR DESCRIPTION
**Pull request by crepererum**
To better gauge how many connections we use and especially if we hit the max connection limit, it would be helpful to actually have some metrics available for the pool usage. This change adds a few basic metrics.

### CODE REVIEW

File: iox_catalog/src/postgres.rs [hunk 1]
```diff
@@ -321,27 +325,207 @@ DO NOTHING;
...
         .min_connections(1)
         .max_connections(options.max_conns)
         .acquire_timeout(options.connect_timeout)
         .idle_timeout(options.idle_timeout)
         .test_before_acquire(true)
+        .before_acquire(move |_c, _meta| {
+            let metrics = metrics_captured_1.clone();
+            let pool_id = Arc::clone(&id_captured_1);
+            Box::pin(async move {
+                metrics.acquire(&pool_id);
```
Comment (mkmik) [inline iox_catalog/src/postgres.rs:517]: I'm confused; who polls this future?

File: iox_catalog/src/postgres.rs [hunk 2]
```diff
@@ -321,27 +325,207 @@ DO NOTHING;
...
+struct PoolMetrics {
+    /// Actual shared state.
+    state: Arc<PoolMetricsInner>,
+}
+
+/// Inner state of [`PoolMetrics`] that is wrapped into an [`Arc`].
+#[derive(Debug, Default)]
+struct PoolMetricsInner {
+    /// Next pool ID.
+    pool_id_gen: AtomicU64,
```
Comment (mkmik) [inline iox_catalog/src/postgres.rs:339]: I wonder if we ever have more than one pool outside tests.

File: iox_catalog/src/postgres.rs (hunk 1 above)
Comment (mkmik) [reply iox_catalog/src/postgres.rs:517]: ah I was confused by github's incorrect symbol analysis. This `acquire` is `PoolMetrics`'s function that doesn't return

**Review by mkmik — APPROVED** [review]

**Comment by mkmik** [comment]
LGTM, although probably unnecessarily complex since we generally have only one pool

File: iox_catalog/src/postgres.rs (hunk 2 above)
Comment (crepererum) [reply iox_catalog/src/postgres.rs:339]: yeah, when you switch over the pool during hotswapping you temporarily have two

**Pull request merged**